recovery/elliptics_recovery/dc_server_send.py

```diff
@@ -0,0 +1,355 @@
+            #bucket.external_sort_by_physical_order()
+            group = bucket.get_group()
+            progress = False
+            for keys in bucket.get_keys(self.ctx.batch_size):
+                progress = True
+                self._server_send(keys, group)
+            bucket.clear()
+        return self.result
+
+    def _server_send(self, keys, group):
+        '''
+        Recovers bunch of newest @keys from replica with group_id == @group to other replicas via server-send.
+        '''
+        log.info("Server-send bucket: source group: {0}, num keys: {1}".format(group, len(keys)))
+        keys_bunch = dict() # remote_groups -> [list of newest keys]
+        for key, key_infos in keys:
+            unprocessed_key_infos = self._get_unprocessed_key_infos(key_infos, group)
+
+            is_first_attempt = len(unprocessed_key_infos) == len(key_infos)
+            if is_first_attempt and self._process_uncommited_keys(key, key_infos):
+                continue
+
+            if not self._can_use_server_send(unprocessed_key_infos):
+                self.buckets.move_to_rest_bucket(key, key_infos)
+                continue
+
+            dest_groups = self._get_dest_groups(unprocessed_key_infos)
+            index = frozenset(dest_groups)
+            if index not in keys_bunch:
+                keys_bunch[index] = []
+            keys_bunch[index].append((key, key_infos))
+
+        self.session.set_groups([group])
+
+        for b in keys_bunch.iteritems():
+            remote_groups = b[0]
+            newest_keys = list()
+            key_infos_map = dict()
+            for key, key_infos in b[1]:
+                log.debug("Prepare server-send key: {0}, group: {1}".format(key, key.group_id))
+                newest_keys.append(key)
+                key_infos_map[str(key)] = key_infos
+
+            timeouted_keys = None
+            for i in range(self.ctx.attempts):
+                #for k in newest_keys:
+                #    result = self.session.lookup(k).get()[0]
+                #    log.debug("LOOKUP2: key: {0}, group: {1}, status: {2}".format(k, result.group_id, result.status))
+                if newest_keys:
+                    log.info("Server-send: group: {0}, remote_groups: {1}, num_keys: {2}".format(group, remote_groups, len(newest_keys)))
+                    iterator = self.session.server_send(newest_keys, 0, remote_groups)
+                    timeouted_keys, corrupted_keys = self._check_server_send_results(iterator, key_infos_map, group)
+                    newest_keys = timeouted_keys
+                    if corrupted_keys:
+                        self._remove_corrupted_keys(corrupted_keys, [group])
+
+            if timeouted_keys:
+                self._on_server_send_timeout(timeouted_keys, key_infos_map, group)
+
+    def _check_server_send_results(self, iterator, key_infos_map, group):
+        '''
+        Check result of remote sending for every key.
+        Returns list of timeouted keys.
```

Comment:
Why do you call them timeouted keys? It looks like it returns a list of failed keys no matter what was the reason